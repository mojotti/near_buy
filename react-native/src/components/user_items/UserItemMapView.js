import React, { Component } from 'react';
import { View } from 'react-native';
import MapView, { Marker } from 'react-native-maps';
import PropTypes from 'prop-types';
import { styles } from '../../static/styles/UserItemDetailsStyles';

export default class UserItemMapView extends Component {
  render() {
    const latitudeDelta =
      Math.abs(this.props.latitude - this.props.currentLocation.latitude) +
      0.004;
    const longitudeDelta =
      Math.abs(this.props.longitude - this.props.currentLocation.longitude) +
      0.0013;
    return (
      <View style={styles.mapContainer}>
        <MapView
          style={styles.map}
          region={{
            latitude: this.props.latitude,
            longitude: this.props.longitude,
            latitudeDelta,
            longitudeDelta,
          }}
        >
          <Marker
            coordinate={{
              longitude: this.props.longitude,
              latitude: this.props.latitude,
            }}
            title={`Item's current location`}
          />
          <Marker
            coordinate={{
              longitude: this.props.currentLocation.longitude,
              latitude: this.props.currentLocation.latitude,
            }}
            title={`Your location`}
          />
        </MapView>
      </View>
    );
  }
}

UserItemMapView.propTypes = {
  longitude: PropTypes.number.isRequired,
  latitude: PropTypes.number.isRequired,
  currentLocation: PropTypes.objectOf(PropTypes.number).isRequired,
};
