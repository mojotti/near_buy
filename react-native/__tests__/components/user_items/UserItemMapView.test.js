import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import UserItemMapView from "../../../src/components/user_items/UserItemMapView";

const CURRENT_LOCATION = {
  latitude: 42.000,
  longitude: 105.00,
};


describe('<UserItemMapView', () => {
  jest.useFakeTimers();

  test('renders correctly', () => {
    const userItemMapView = shallow(
      <UserItemMapView
        longitude={25.000}
        latitude={64.000}
        currentLocation={CURRENT_LOCATION}
      />,
    );
    expect(userItemMapView).toMatchSnapshot();
  });

  test('gets center location', () => {
    const userItemMapView = shallow(
      <UserItemMapView
        longitude={25.000}
        latitude={64.000}
        currentLocation={CURRENT_LOCATION}
      />,
    );

    const expectedCenterLocation = { centerLatitude: 53, centerLongitude: 65 };
    const centerLocation = userItemMapView.instance().getCenterLocation();

    expect(centerLocation).toEqual(expectedCenterLocation);
  });

  test('gets location deltas', () => {
    const userItemMapView = shallow(
      <UserItemMapView
        longitude={25.000}
        latitude={64.000}
        currentLocation={CURRENT_LOCATION}
      />,
    );

    const expectedDeltas = { latitudeDelta: 22.006, longitudeDelta: 80.0017 };
    const deltas = userItemMapView.instance().getLocationDeltas();

    expect(deltas).toEqual(expectedDeltas);
  });

  test('animates to new location when location changes', () => {
    const userItemMapView = shallow(
      <UserItemMapView
        longitude={25.000}
        latitude={64.000}
        currentLocation={CURRENT_LOCATION}
      />,
    );

    const instance = userItemMapView.instance();
    const animateSpy = jest.spyOn(instance, 'animateItemToNewLocation');
    expect(animateSpy.mock.calls.length).toEqual(0);

    instance.componentDidUpdate(CURRENT_LOCATION);
    expect(animateSpy.mock.calls.length).toEqual(1);
  });

  test('does not animate to new location when location stays the same', () => {
    const userItemMapView = shallow(
      <UserItemMapView
        longitude={25.000}
        latitude={64.000}
        currentLocation={CURRENT_LOCATION}
      />,
    );

    const instance = userItemMapView.instance();
    const animateSpy = jest.spyOn(instance, 'animateItemToNewLocation');
    expect(animateSpy.mock.calls.length).toEqual(0);

    instance.componentDidUpdate({ latitude: 64, longitude: 25 });
    expect(animateSpy.mock.calls.length).toEqual(0);
  });
});
