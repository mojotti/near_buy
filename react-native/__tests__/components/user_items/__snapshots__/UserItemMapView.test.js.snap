// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserItemMapView renders correctly 1`] = `
<Component
  style={
    Object {
      "alignItems": "center",
      "height": 600.3000000000001,
      "marginBottom": 15,
      "marginTop": 15,
      "width": 750,
    }
  }
>
  <MapView
    initialRegion={
      Object {
        "latitude": 53,
        "latitudeDelta": 22.006,
        "longitude": 65,
        "longitudeDelta": 80.0017,
      }
    }
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <AnimatedComponent
      coordinate={
        Object {
          "latitude": 64,
          "latitudeDelta": NaN,
          "longitude": 25,
          "longitudeDelta": NaN,
        }
      }
      pinColor="blue"
      title="Item's current location"
    />
    <AnimatedComponent
      coordinate={
        Object {
          "latitude": 42,
          "longitude": 105,
        }
      }
      title="Your location"
    />
  </MapView>
</Component>
`;
